### DISCUSSION TITLE
Automating molecular animations using PyMOL and Blender APIs

### DISCUSSION THREAD

**Discussion by Orbitaliant** [BradyAJohnston/MolecularNodes#540] [Ideas]
Hellooo,

I'm thinking about creating molecular animation pipelines using Python, PyMOL, and Blender! However, I'm not sure how feasible it is to use molecular nodes through Blender's API.

I look forward to your thoughts,

Islam

**Comment by BradyAJohnston** [comment]
hi mate! It is certainly possible to use Molecular Nodes through the python API inside of Blender - but the API for MN should be considered not stable at the moment. There has been some considerable overhaul occurring to how the data is processed under the hood and how potential users could interact with it.

First and foremost MN is designed to be a GUI application that users interact with via pointing and clicking inside of Blender. Everything has been created mostly with that in mind, and secondarily the potential to script with it. 

I would like to improve the potential for scripting with it, and there has been considerable effort put towards this goal for the 4.2 release (#501) release. 

I welcome discussion on the topic and had previously opened an issue for discussion on here #359.

**Comment by BradyAJohnston** [comment]
I'm also not sure what role PyMOL would be playing here - if Molecular Nodes is generating the 3D models from the data?

**Reply by Orbitaliant** [reply]
Hi Brady,

PyMOL offers many features for manipulating structural data (such as structural alignments or constructing biological assemblies) while maintaining computational chemistry parameters. Therefore, it allows the user to perform many calculations before proceeding to visualization, which would be convenient for workflows (e.g. using Jupyter). Through Blender's API, creating molecular modeling workflows culminating with awesome 3D models using Blender's graphics would be very convenient.

What do you think? I might be overthinking haha

Many thanks mate

**Reply by rbdavid** [reply]
MDAnalysis might be better suited for scripting complex comp chem analyses than working within Pymol's scripting environment. That said, Pymol does have some nice structural alignment methods encoded natively so I can't fault you for using it within your workflow. Along that line, I'm a bit stuck on a project of converting structural alignment functions within the USalign software to python handles so that I can implement them natively within python scripts (such as those used in blender and MN).

On a different front, Brady is doing a bunch of work to enable users to use a MDAnalysis Universe within a blender project; I think #535 is a good example of this.

**Reply by Orbitaliant** [reply]
Hi Russell,

Many thanks for your recommendation, I'll try MDAnalysis in my current workflow.

Have you considered CEAlign instead of USalign? CEAlign has been working more robustly for me.

Let me know if I can be of further assistance,

All the best

**Reply by rbdavid** [reply]
As I understand it, cealign is an alignment method in Pymol. Or is there a separate code base that allows you to use it outside of pymol? I've used cealign in pymol before but I quickly moved away from pymol so didn't test cealign further.  

Also, cealign and all of pymol's alignment methods are sequence-order dependent; they do a sequence alignment to get an initial prediction for the alignment. I won't go on a long spiel about sequence-order-dependent and -independent methods but USalign is a C++ code that has functionality to do either as well as align nucleic oligomers and a bunch of other things. USalign is pretty damn good in its diversity of alignment methods it can do. Its up to the user to decide which method will return the best alignment though. USalign's ability to do so many different alignments is a double edged sword; there are many many many parameters for a user to figure out.

**Reply by Orbitaliant** [reply]
USalign is pretty good, don't get me wrong, I just thought cealign might be a quick implement if that's what you need for your current project.

You can also run cealign using biopython if you prefer not to use pymol:

https://biopython.org/docs/dev/api/Bio.PDB.cealign.html

Best of luck
